4_predict/src/train_models.R

```diff
@@ -555,6 +555,228 @@ filter_rows_date <- function(attrs, start_date){
 
 
 #SHAP values
+setup_shap_data <- function(data, seasonal = TRUE, physio = TRUE, lulc = TRUE,
+                            lulc_prop = 0.75, lulc_data = NULL, physio_data = NULL){
+  #' @description prepares data for SHAP computation based on the provided split
+  #'
+  #' @param data the dataset with predictor attributes
+  #' @param seasonal logical for splitting data by water year season
+  #' @param physio logical for splitting data by physiographic region
+  #' @param lulc logical for splitting data by lulc, as defined into 2 groups:
+  #' high urban land cover and high forest land cover
+  #' @param lulc_prop threshold defining "high" proportion land cover on [0,1]
+  #' @param lulc_data land cover data for total upstream urban and forest. Must
+  #' also have the PRMS_segid and Date columns
+  #' @param physio_data physiographic region data. 
+  #' Must also have the PRMS_segid and Date columns
+  #' 
+  #' @return Returns a list with the different data splits
+  
+  data_lst <- list()
+  data_lst_names <- c()
+  
+  if(seasonal){
+    #split data by water year season (4 elements in list)
+    data$months <- lubridate::month(data$Date)
+    OND <- filter(data, months %in% c(10,11,12)) %>%
+      select(PRMS_segid, Date)
+    JFM <- filter(data, months %in% c(1,2,3)) %>%
+      select(PRMS_segid, Date)
+    AMJ <- filter(data, months %in% c(4,5,6)) %>%
+      select(PRMS_segid, Date)
+    JAS <- filter(data, months %in% c(7,8,9)) %>%
+      select(PRMS_segid, Date)
+    
+    data_lst <- c(data_lst, list(OND), list(JFM), list(AMJ), list(JAS))
+    #names for list elements
+    data_lst_names <- c(data_lst_names, paste0('seas_', c('OND', 'JFM', 'AMJ', 'JAS')))
+    
+    if(lulc){
+      #split each season by lulc
+      ind_seasons <- grep('^seas_', data_lst_names)
+      for(i in 1:4){
```

Comment:
To make the for loops more robust to unexpected values of `ind_seasons` you could consider adding a check to this function that the length of `ind_seasons` is 4, e.g. `stopifnot(length(ind_seasons) == 4)` or just loop by `seq(ind_seasons)`.   This is a minor suggestion though, so I'll leave that up to you.

Reply:
As programmed, I think `data_lst_names` is guaranteed to have 4 elements. Some of the elements might be a data frame with 0 rows, but the element will exist. But yeah, it's possible that we later edit the function to rearrange the order of these function's steps and then this loop could give unexpected results. I'll change this